resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/KubernetesClusterManager.scala

```diff
@@ -56,17 +56,43 @@ private[spark] class KubernetesClusterManager extends ExternalClusterManager wit
       Some(new File(Config.KUBERNETES_SERVICE_ACCOUNT_TOKEN_PATH)),
       Some(new File(Config.KUBERNETES_SERVICE_ACCOUNT_CA_CRT_PATH)))
 
-    val allocatorExecutor = ThreadUtils
-      .newDaemonSingleThreadScheduledExecutor("kubernetes-pod-allocator")
     val requestExecutorsService = ThreadUtils.newDaemonCachedThreadPool(
       "kubernetes-executor-requests")
+
+    val bufferEventsExecutor = ThreadUtils
+      .newDaemonSingleThreadScheduledExecutor("kubernetes-executor-pods-event-buffer")
+    val executeEventSubscribersExecutor = ThreadUtils
+      .newDaemonCachedThreadPool("kubernetes-executor-pods-event-handlers")
+    val eventQueue = new ExecutorPodsEventQueueImpl(
+      bufferEventsExecutor, executeEventSubscribersExecutor)
```

Comment:
You also need to shutdown those executors?

Reply:
Given that we now have several executors to shutdown, can we shutdown them all in as few places as possible, e.g., in `KubernetesClusterSchedulerBackend.stop`? For example, you can shutdown `requestExecutorsService` and `eventsPollingExecutor` in `KubernetesClusterSchedulerBackend.stop`. Shutdown of `bufferEventsExecutor` and `executeEventSubscribersExecutor` probably need to handled by `ExecutorPodsEventQueueImpl` in its `stopProcessingEvents` method. Ideally they should be shutdown in the same place where they are created, but it's not doable in this case because `KubernetesClusterManager` doesn't have a stop method.

Reply:
Think we'd want to then have all the `start` methods take the scheduled executor service as an argument, then the scheduler backend owns shutting them down. But I don't think that's the right abstraction - the subcomponents own the thread pools and know how these thread pools are being used, when tasks are being submitted to them, etc. I think it's fine to shut down these thread pools from the subcomponents.

Reply:
OK, I'm fine with that.

Reply:
One way to share this logic would be to have a shared trait for all the things that have thread pools to stop, and then use the trait to stop everything. e.g.

```
trait HasThreadPools {

  protected def threadPools(): Iterable[Executor]
  
  def stop(): Unit = threadPools().foreach(_.shutdownNow())
}

class MyThreadPoolsDependentClass {

  override def threadPools(): Iterable[Executor] = { // All the executors I want to shut down }
  
  override def stop(): Unit = {
    // Stop everything else
    super.stop()
  }
}
```

Not sure if it's worth the extra abstraction to do this though.

Reply:
One way to share this logic would be to have a shared trait for all the things that have thread pools to stop, and then use the trait to stop everything. e.g.

```
trait HasThreadPools {

  protected def threadPools(): Iterable[Executor]
  
  def stop(): Unit = threadPools().foreach(_.shutdownNow())
}

class MyThreadPoolsDependentClass {

  override def threadPools(): Iterable[Executor] = { // All the executors I want to shut down }
  
  override def stop(): Unit = {
    // Stop everything else
    super.stop()
  }
}
```

Not sure if it's worth the extra abstraction to do this though.

Reply:
I don't think it's worth it. I'm fine shutting them down inside the subcomponents as long as the contract that each thread-pool is only used by a single owning component holds.

Reply:
Done